### PR TITLE
ISSUE-125 Handle arrays with to/fromJSONStrategy properly

### LINKED ISSUE #125: [BUG] - Using strategies on Array types runs the strategy on array elements instead of the whole array
**Describe the bug** 
When serializing there is a check for `Array.isArray(value)` which, if true, runs the serialization on the individual elements of the array and returns an array of the serialized objects. This is fine in most cases but if you want a custom strategy to serialize an array into a string for example, the behaviour doesnt work, since now you will get an array of the serialized strings. When using strategies at least, the user should be allowed to define the strategy to run on the whole object no matter if it is an array or not. 

**To Reproduce**
```
// Lets say we have an object with a property that is an array of objects that we want to serialize into one single string
@SerializeProperty({ toJSONStrategy: objArraySerialize, fromJSONStrategy: objArrayDeserialize })
public myProp:{id: number}[];
...

// We expect the strategy will be called with the full array and we can serialize it in one shot. 
const objArraySerialize = (items: {id: number}[]): string {
  return JSON.stringify(items.map(el => el.id));
}

```

In the above what will actually happen is that our `objArraySerialize` will get called for every element in the array instead of the full array.

**Expected behaviour** 
The user should be able to define custom strategies for arrays that use the custom strategy for the entire array rather than just the elements on the array

### PR DESCRIPTION
**Pull request by khan-saim**
**Description**

Issue fixed: [125](https://github.com/GameBridgeAI/ts_serialize/issues/125)

Here we change the to/fromJSONStrategy handling when the property is an array. If a strategy is provided, it is applied on the full array rather than individually on the elements of the array. 

**Things to look at**

- [ ] Test coverage
- [ ] Code Style
- [ ] Documentation (`README.md`, `doc/`, `examples/`, etc..)

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -1 +1,6 @@
-{ "deno.enable": true }
+{
+    "deno.enable": true,
+    "deno.lint": true,
+    "deno.unstable": false,
+    "deno.path": "/Users/saim.khan/.deno/bin/deno"
+}
```
Comment (khan-saim) [inline .vscode/settings.json]: oh wait what, why did this sneak in 😭

File: .gitignore [hunk 2]
```diff
@@ -7,4 +7,5 @@ examples/node/*
 !examples/node/README.md
 coverage
 coverage.lcov
-.DS_Store
\ No newline at end of file
+.DS_Store
+.vscode
```
Comment (hardy925) [inline .gitignore:11]: We want to keep  `.vscode`, what issues are you seeing?

File: .vscode/settings.json [hunk 3]
```diff
@@ -1 +0,0 @@
-{ "deno.enable": true }
```
Comment (hardy925) [inline .vscode/settings.json:1]: we want this guy too

File: serializable.ts [hunk 4]
```diff
@@ -152,12 +152,19 @@ export function toPojo(
     }
 
     if (Array.isArray(value)) {
+      let notSerializableElements = false;
       record[serializedKey] = value.map((item: any) => {
         if (item instanceof Serializable) {
           return toPojo(item);
         }
```
Comment (hardy925) [inline serializable.ts:159]: so. I don't think any of this stuff is really needed now

File: serializable.ts [hunk 5]
```diff
@@ -152,12 +152,19 @@ export function toPojo(
...
-        return toJSONStrategy(item);
+
+        // if item is not Serializable we will apply the toJSONStrategy on the whole array so return items as is
+        notSerializableElements = true;
```
Comment (hardy925) [inline serializable.ts:162]: not sure if we need this guy too

File: serializable.ts [hunk 6]
```diff
@@ -152,12 +152,19 @@ export function toPojo(
...
+        return item;
       });
+      if(notSerializableElements) {
+        record[serializedKey] = toJSONStrategy(value);
+      }
```
Comment (hardy925) [inline serializable.ts:167]: this can be simplified now, and I don't think we need the check

File: serializable.ts [hunk 7]
```diff
@@ -181,7 +188,7 @@ function fromJSON<T>(
   for (const [key, value] of Object.entries(_json)) {
     const {
       propertyKey,
-      fromJSONStrategy = fromJSONDefault,
+      fromJSONStrategy,
```
Comment (hardy925) [inline serializable.ts:191]: I think we'll want to revert this change

File: serializable.ts [hunk 8]
```diff
@@ -190,9 +197,16 @@ function fromJSON<T>(
...
-    accumulator[propertyKey as keyof T] = Array.isArray(value)
-      ? value.map((v) => fromJSONStrategy(v))
-      : fromJSONStrategy(value as JSONValue);
+    // if strategy is not provided we apply the default on the value/elements of the array
+    if (!fromJSONStrategy) {
+      const defaultFromJSONStrategy = fromJSONDefault;
+      accumulator[propertyKey as keyof T] = Array.isArray(value)
+      ? value.map((v) => defaultFromJSONStrategy(v)) : defaultFromJSONStrategy(value as JSONValue)
+      continue;
+    }
```
Comment (hardy925) [inline serializable.ts:206]: this can most likely be removed now too

File: serializable.ts [hunk 9]
```diff
@@ -190,9 +197,16 @@ function fromJSON<T>(
...
+
+    // if a strategy is provided we apply the strategy on the entire array
+    accumulator[propertyKey as keyof T] = fromJSONStrategy(value as JSONValue);
```
Comment (hardy925) [inline serializable.ts:209]: this is basically it now because we want to pass the direct value through the strat array or not, and we and guarantee the default set by the destructuring.

File: serialize_property_test.ts [hunk 10]
```diff
@@ -560,3 +560,14 @@ test({
...
+test({
+  name: "custom toJSONStrategy with an array of data",
+  fn() {
+    class Test extends Serializable {
+      @SerializeProperty({ toJSONStrategy: (v) => v.join(".") })
+      public test_property = [0,1];
+    }
+    assertEquals(new Test().toJSON(), `{"test_property":"0.1"}`);
+  },
+});
```
Comment (hardy925) [inline serialize_property_test.ts:573]: this is a good test, but I think we can also remove some other tests

File: strategy/from_json/to_serializable_test.ts [hunk 11]
```diff
@@ -104,7 +104,7 @@ test({
     }
     class Test extends Serializable {
       @SerializeProperty({
-        fromJSONStrategy: (v) => new OtherClass().fromJSON(v),
+        fromJSONStrategy: (v) => (v as JSONArray).map(el => new OtherClass().fromJSON(el)),
```
Comment (hardy925) [inline strategy/from_json/to_serializable_test.ts:107]: so you can type this at the param level and it should apply to the function scope

**Review by hardy925 — CHANGES_REQUESTED** [review]
Hey @khan-saim :wave:

Thanks for opening a PR, I did a quick skim of the code and I think we still have some changes to make. For the most part, I believe we would be removing more code than adding.

Let me know if you have questions

Cheers,

**Comment by shardyMBAI** [comment]
Hey @khan-saim, Thanks again for the PR, but I am going to close it as we've recently pushed some rather larger changes, while doing these changes I also took care of this issue with #131

**Comment by hardy925** [comment]
Hey @khan-saim, Thanks again for the PR, but I am going to close it as we've recently pushed some rather larger changes, while doing these changes I also took care of this issue with #131
